src/transform-sdk/rust/core-types/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::SystemTime;
+
+/// An event generated after a write event within the broker.
+#[derive(Debug)]
+pub struct WriteEvent<'a> {
+    /// The record for which the event was generated for.
+    pub record: BorrowedRecord<'a>,
+}
+
+/// A zero-copy record header.
+#[derive(Debug)]
+pub struct BorrowedHeader<'a> {
+    key: &'a [u8],
+    value: Option<&'a [u8]>,
+}
+
+impl<'a> BorrowedHeader<'a> {
+    /// Create a new header without any copies.
+    pub fn new(key: &'a [u8], value: Option<&'a [u8]>) -> Self {
+        Self { key, value }
+    }
+
+    /// Returns the header's key or `None` if there is no key.
+    pub fn key(&self) -> &[u8] {
+        self.key
+    }
+
+    /// Returns the header's value or `None` if there is no value.
+    pub fn value(&self) -> Option<&[u8]> {
+        self.value
+    }
+
+    /// Clones this header obtaining ownership of the clone.
+    pub fn to_owned(&self) -> RecordHeader {
+        RecordHeader::new(self.key.to_owned(), self.value.map(|b| b.to_owned()))
+    }
```

Comment:
I believe you'll want to implement the [`ToOwned`](https://doc.rust-lang.org/std/borrow/trait.ToOwned.html) trait for this

Reply:
I believe this provides support to the [`Borrowed`](https://doc.rust-lang.org/std/borrow/trait.Borrow.html) trait for free

Reply:
So I tried this and didn't get it to work, here is the stackoverflow I found that pointed me to just make the non trait version: https://stackoverflow.com/questions/72105604/implement-toowned-for-user-defined-types

Reply:
gotcha, thanks for the info!